README
OmniPotent v1.2.2
=====================================================================
Copyright (C) 2014 Lock of War, Developmental (Pty) Ltd

RAID STYLE UNIT FRAMES in ARENA
ENEMY AND PARTY FRAME GROUPS
POKEMON NAMES for RUSSIAN TARGET CALLING

Supports Legion v7.2.0

Download MrTarget for BATTLEGROUNDS
http://www.curse.com/addons/wow/mrtarget

CHECK OUT THE ADDON IN ACTION WITH SACHIE
http://youtube.com/c/lockofwar

SEND BUGS or FEATURE REQUESTS TO
http://twitter.com/renevatium OR hello@renevatium.com

Right-click drag to MOVE (only while options panel is open)
Unit frame orientation is automatic based on screen position
Objectives frame will be collapsed during battle
To avoid taint, options that alter the UI are disabled during battle

MORE FEEDBACK PLEASE!
We want to hear you ideas/issues

SLASH COMMANDS
---------------------------------------------------------------------
/omnipotent, /omni, /op - Open interface options panel

LICENSE
---------------------------------------------------------------------
This Work is provided under the Creative Commons Attribution-NonCommercial-NoDerivatives 4.0
International Public License. For more information see LICENSE.txt included with this distribution
or visit https://creativecommons.org/licenses/by-nc-nd/4.0/ for the full text.

CHANGELOG
OmniPotent v1.2.2
=====================================================================
Copyright (C) 2014 Lock of War, Developmental (Pty) Ltd

Legion v7.2.0
---------------------------------------------------------------------
30/03/2017
v1.2.2
Updated to support patch 7.2.0 (70200)

Legion v7.1.5
---------------------------------------------------------------------
18/02/2017
v1.2.1
Fixed for updated GetSpecializationInfoByID

15/01/2017
v1.2.0
Updated to support patch 7.1.5 (70100)
Fixed for updated GetSpecializationInfoForClassID

Legion v7.1.0
---------------------------------------------------------------------
28/10/2016
v1.1.0
Updated to support patch 7.1.0 (70100)

Legion v7.0.3
---------------------------------------------------------------------
16/04/2016
v1.0.9
Full fade when out of range
Fixed range logic

21/07/2016
v1.0.8
Updated to support patch 7.0.3 (70000)

Warlords of Draenor v6.2.4
---------------------------------------------------------------------
16/04/2016
v1.0.7
Updated to support patch 6.2.4 (60200)
Fixed error from GetSpecializationInfo

Warlords of Draenor v6.2.3
---------------------------------------------------------------------
18/11/2015
v1.0.6
Updated to support patch 6.2.3 (60200)
Units order consistent with native frames

Warlords of Draenor v6.2.2
---------------------------------------------------------------------
07/09/2015
v1.0.5
Updated to support patch 6.2.2 (60200)

Warlords of Draenor v6.2.1
---------------------------------------------------------------------
27/08/2015
v1.0.4
Fixed loading sequence
Fixed right to left display

21/08/2015
v1.0.3
Updated to support patch 6.2.1 (60200)
Fixed range tracking
Fixed objectives frame toggle

Warlords of Draenor v6.2.0
---------------------------------------------------------------------
18/08/2015
v1.0.2
Auras now display DEBUFFS only
Fixed unload process
Party frame option

10/08/2015
v1.0.1
Party frames

02/08/2015
v1.0.0
First build

OmniPotentUnit.lua
-- OmniPotentUnit
-- =====================================================================
-- Copyright (C) 2014 Lock of War, Developmental (Pty) Ltd
--

local INSPECT = false;
local POWER_BAR_COLORS = {
  MANA={ r=0.00,g=0.00,b=1.00 },
  RAGE={ r=1.00,g=0.00,b=0.00 },
  FOCUS={ r=1.00,g=0.50,b=0.25 },
  ENERGY={ r=1.00,g=1.00,b=0.00 },
  CHI={ r=0.71,g=1.0,b=0.92 },
  RUNES={ r=0.50,g=0.50,b=0.50 }
};

local LFGRoleTexCoords = { TANK={ 0.5,0.75,0,1 }, DAMAGER={ 0.25,0.5,0,1 }, HEALER={ 0.75,1,0,1 }};
local function GetTexCoordsForRole(role, borderless)
  role = role or 'DAMAGER';
  local c = borderless and LFGRoleTexCoords[role] or {GetTexCoordsForRoleSmallCircle(role)};
  return unpack(c);
end

OmniPotentUnit = {
  update=0,
  parent=0,
  dead=false,
  test=false,
  name=nil,
  icon=nil,
  display=nil,
  unit=nil,
  frame=nil,
  spec=nil,
  class=nil,
  role=nil,
  health=1,
  healthMax=1,
  power=1,
  powerMax=1,
  targeted='',
  range=nil,
  track=nil,
  last_update=0,
  update_targeted=true,
  enemy=false,
  auras=nil
};

OmniPotentUnit.__index = OmniPotentUnit;

function OmniPotentUnit:New(parent, count)
  local this = setmetatable({}, OmniPotentUnit);
  this.parent = parent;
  this.frame = CreateFrame('Button', parent.frame:GetName()..'OmniPotentUnit'..count, parent.frame, 'OmniPotentUnitTemplate');
  this.frame:SetScript('OnEvent', function(frame, ...) this:OnEvent(...); end);
  this.frame:SetScript('OnUpdate', function(frame, ...) this:OnUpdate(...); end);
  this.frame:SetScript('OnEnter', function(frame, ...) this:OnEnter(...); end);
  this.frame:SetScript('OnLeave', function(frame, ...) this:OnLeave(...); end);
  this.frame:SetScript('OnDragStart', function(frame, ...) parent:OnDragStart(...); end);
  this.frame:SetScript('OnDragStop', function(frame, ...) parent:OnDragStop(...); end);
  this.frame.UPDATE_TARGETED:SetScript('OnUpdate', function(frame, ...) this:UpdateTargeted(...); end);
  this.frame:EnableMouse(true);
  this.frame:RegisterForDrag('RightButton');
  this.frame:RegisterForClicks('LeftButtonUp', 'RightButtonUp');
  this.frame:SetAttribute('type1', 'macro');
  this.frame:SetAttribute('type2', 'macro');
  this.frame:SetAttribute('macrotext1', '');
  this.frame:SetAttribute('macrotext2', '');
  this.auras = OmniPotentAuras:New(this);
  this.track = OmniPotentRange:New(this);
  return this;
end

function OmniPotentUnit:SetUnit(unit, name, display, class, spec, role, icon, test)
  self.unit = unit;
  self.frame.unit = unit;
  self.test = test;
  if name then
    self.name = name;
    self.display = display;
    self.spec = spec;
    self.class = class;
    self.role = role;
    self.icon = icon;
    self:RegisterEvents();
    self:SetFrameStyle();
    self.frame:Show();
  else
    self.frame:RegisterEvent('INSPECT_READY');
    if not INSPECT then
      INSPECT = unit;
      NotifyInspect(unit);
    end
  end
end

function OmniPotentUnit:SetName(name, display)
  self.name = name;
  self.display = display;
end

function OmniPotentUnit:UnsetUnit()
  self.name = nil;
  self.display = nil;
  self.class = nil;
  self.spec = nil;
  self.role = nil;
  self.unit = nil;
  self.frame.unit = nil;
  self.dead = false;
  self.test = false;
  self.frame.SPEC:SetText('');
  self.frame.SPEC_ICON:SetTexture(nil);
  self.frame.NAME:SetText('');
  self.health = 1;
  self.healthMax = 1;
  self.power = 1;
  self.powerMax = 1;
  self:UnregisterEvents();
  self:Hide();
end

function OmniPotentUnit:InspectReady(guid)
  if UnitGUID(self.unit) == guid then
    self.frame:UnregisterEvent('INSPECT_READY');
    local name = GetUnitName(self.unit, true) or 'Unknown';
    local display = self.parent:GetDisplayName(name);
    local id = GetInspectSpecialization(self.unit);
    local id, spec, description, icon, role, class = GetSpecializationInfoByID(id);
    if spec then
      self:SetUnit(self.unit, name, display, class, spec, role, icon, test);
      INSPECT = false;
    end
  elseif not INSPECT then
    INSPECT = self.unit;
    NotifyInspect(self.unit);
  end
end

function OmniPotentUnit:PlayerRegenEnabled()
  self.frame:UnregisterEvent('PLAYER_REGEN_ENABLED');
  self.frame:SetAttribute('macrotext1', nil);
  self.frame:SetAttribute('macrotext2', nil);
  self.frame:Hide();
end

function OmniPotentUnit:Hide()
  if InCombatLockdown() then
    self.frame:RegisterEvent('PLAYER_REGEN_ENABLED');
  else
    self.frame:Hide();
  end
end

function OmniPotentUnit:Destroy()
  self:UnsetUnit();
  self.auras:Destroy();
end

function OmniPotentUnit:UnitHealthColor()
  local color = RAID_CLASS_COLORS[self.class];
  self.frame.HEALTH_BAR:SetStatusBarColor(color.r, color.g, color.b);
  self.frame.HEALTH_BAR.r, self.frame.HEALTH_BAR.g, self.frame.HEALTH_BAR.b = color.r, color.g, color.b;
end

function OmniPotentUnit:UnitPowerColor()
  local powerType, powerToken = UnitPowerType(self.name);
  local color = POWER_BAR_COLORS[powerToken] or POWER_BAR_COLORS.MANA;
  self.frame.POWER_BAR:SetStatusBarColor(color.r, color.g, color.b);
end

function OmniPotentUnit:SetAlpha(alpha)
  self:UnitHealthColor(alpha);
  self:UnitPowerColor(alpha);
  self.frame.SPEC_ICON:SetAlpha(alpha);
  self.frame:SetAlpha(alpha);
end

function OmniPotentUnit:UnitUpdate()
  if self.unit then
    self.last_update = GetTime();
    self.enemy = UnitIsEnemy(self.unit, 'player');
    self.health = UnitHealth(self.unit);
    self.healthMax = UnitHealthMax(self.unit);
    self.power = UnitPower(self.unit);
    self.powerMax = UnitPowerMax(self.unit);
    if UnitIsDeadOrGhost(self.unit) or self.health == 0 then
      self.power = 0;
      self.range = nil;
      self.dead = true;
      self.auras:Destroy();
    else
      self.dead = false;
      self.auras:UnitAura(self.unit);
    end
  end
end

function OmniPotentUnit:UnitCheck(unit)
  if UnitIsUnit(self.unit, unit) then
    self:UnitUpdate();
  end
end

function OmniPotentUnit:UpdateTargeted(unit)
  if OmniPotent.OPTIONS.TARGETED then
    if self.update_targeted then
      self.targeted = 0;
      self.update_targeted = false;
      for i=1, GetNumGroupMembers() do
        if UnitIsUnit('raid'..i..'target', self.unit) then
          self.targeted = self.targeted+1;
        end
      end
      self.targeted = self.targeted>0 and self.targeted or '';
      self.update_targeted = true;
    end
  end
end

function OmniPotentUnit:OnUpdate(time)
  self.update = self.update + time;
  if self.update < 0.2 then
    return;
  end
  self.update = 0;
  self:UpdateDisplay();
end

function OmniPotentUnit:UpdateDisplay()
  self.frame.NAME:SetText(self.display);
  self.frame.SPEC:SetText(self.spec);
  self.frame.SPEC_ICON:SetTexture(self.icon);
  self.frame.SPEC_ICON:SetAlpha(1);
  self.frame.HEALTH_BAR:SetMinMaxValues(0, self.healthMax);
  self.frame.HEALTH_BAR:SetValue(self.health);
  self.frame.POWER_BAR:SetMinMaxValues(0, self.powerMax);
  self.frame.POWER_BAR:SetValue(self.power);
  self.frame.TARGETED:SetText(self.targeted);
  self:UnitHealthColor();
  self:UnitPowerColor();
  self:ResetTargetMacro();
  if OmniPotent.OPTIONS.RANGE and self.range == nil then
    self:SetAlpha(0.5);
  else
    self:SetAlpha(1);
  end
end

function OmniPotentUnit:ResetTargetMacro()
  if not InCombatLockdown() then
    self.frame:SetAttribute('macrotext1', '/targetexact '..self.unit);
    self.frame:SetAttribute('macrotext2', '/targetexact '..self.unit..'\n/focus\n/targetlasttarget');
  end
end

function OmniPotentUnit:OnEnter() self.frame.HOVER:Show(); end
function OmniPotentUnit:OnLeave() self.frame.HOVER:Hide(); end

function OmniPotentUnit:OnEvent(event, unit, x, y, z)
  if event == 'INSPECT_READY' then self:InspectReady(unit);
  elseif event == 'UNIT_HEALTH_FREQUENT' then self:UnitCheck(unit);
  elseif event == 'UNIT_COMBAT' then self:UnitCheck(unit);
  elseif event == 'UNIT_TARGET' then self:UnitCheck(unit);
  elseif event == 'UPDATE_MOUSEOVER_UNIT' then self:UnitCheck('mouseover');
  elseif event == 'PLAYER_REGEN_ENABLED' then self:PlayerRegenEnabled();
  end
end

function OmniPotentUnit:RegisterEvents()
  self.frame:RegisterEvent('UNIT_HEALTH_FREQUENT');
  self.frame:RegisterEvent('UPDATE_MOUSEOVER_UNIT');
  self.frame:RegisterEvent('UNIT_COMBAT');
  self.frame:RegisterEvent('UNIT_TARGET');
end

function OmniPotentUnit:UnregisterEvents()
  self.frame:UnregisterEvent('INSPECT_READY');
  self.frame:UnregisterEvent('UNIT_HEALTH_FREQUENT');
  self.frame:UnregisterEvent('UPDATE_MOUSEOVER_UNIT');
  self.frame:UnregisterEvent('UNIT_COMBAT');
  self.frame:UnregisterEvent('UNIT_TARGET');
end

function OmniPotentUnit:SetFrameStyle()
  if OmniPotent.OPTIONS.BORDERLESS then self:SetStyleBorderless();
  else self:SetStyleDefault();
  end
end

function OmniPotentUnit:SetStyleDefault()
  self.frame:EnableDrawLayer('BORDER');
  self.frame.NAME:SetFontObject("GameFontHighlight");
  self.frame.TARGETED:SetFontObject("TextStatusBarTextRed");
  self.frame.HEALTH_BAR:ClearAllPoints();
  self.frame.HEALTH_BAR:SetPoint('TOPLEFT', self.frame, 'TOPLEFT', 1, -1);
  if OmniPotent.OPTIONS.POWER then
    self.frame.POWER_BAR:ClearAllPoints();
    self.frame.HEALTH_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', -1, 10);
    self.frame.POWER_BAR:SetPoint('TOPLEFT', self.frame.HEALTH_BAR, 'BOTTOMLEFT', 0, -2);
    self.frame.POWER_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', -1, 0);
    self.frame.POWER_BAR:Show();
    self.frame.horizDivider:Show();
  else
    self.frame.HEALTH_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', -1, 2);
    self.frame.POWER_BAR:Hide();
    self.frame.horizDivider:Hide();
  end
  self.frame.ROLE_ICON:ClearAllPoints();
  self.frame.ROLE_ICON:SetTexture('Interface\\LFGFrame\\UI-LFG-ICON-PORTRAITROLES');
  self.frame.ROLE_ICON:SetPoint('TOPLEFT', self.frame, 'TOPLEFT', 2.5, -2.5);
  self.frame.ROLE_ICON:SetTexCoord(GetTexCoordsForRole(self.role, false));
  self.frame.ROLE_ICON:Show();
  self.frame.SPEC:Show();
  self.frame.SPEC_ICON:Hide();
end

function OmniPotentUnit:SetStyleBorderless()
  self.frame:DisableDrawLayer('BORDER');
  self.frame.NAME:SetFontObject("GameFontHighlightBorderless");
  self.frame.TARGETED:SetFontObject("TextStatusBarTextRedBorderless");
  self.frame.HEALTH_BAR:ClearAllPoints();
  self.frame.HEALTH_BAR:SetPoint('TOPLEFT', self.frame, 'TOPLEFT', 0, 0);
  if OmniPotent.OPTIONS.POWER then
    self.frame.POWER_BAR:ClearAllPoints();
    self.frame.HEALTH_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', 0, 15);
    self.frame.POWER_BAR:SetPoint('TOPLEFT', self.frame.HEALTH_BAR, 'BOTTOMLEFT', 0, -1);
    self.frame.POWER_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', 0, 1);
    self.frame.POWER_BAR:Show();
    self.frame.horizDivider:Show();
  else
    self.frame.HEALTH_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', 0, 1);
    self.frame.POWER_BAR:Hide();
    self.frame.horizDivider:Hide();
  end
  self.frame.POWER_BAR:ClearAllPoints();
  self.frame.POWER_BAR:SetPoint('TOPLEFT', self.frame.HEALTH_BAR, 'BOTTOMLEFT', 0, -1);
  self.frame.POWER_BAR:SetPoint('BOTTOMRIGHT', self.frame, 'BOTTOMRIGHT', 0, 1);
  self.frame.POWER_BAR:Show();
  self.frame.ROLE_ICON:ClearAllPoints();
  self.frame.ROLE_ICON:SetTexture('Interface\\LFGFrame\\LFGRole');
  self.frame.ROLE_ICON:SetPoint('TOPLEFT', self.frame, 'TOPLEFT', 2.5, -3);
  self.frame.ROLE_ICON:SetTexCoord(GetTexCoordsForRole(self.role, true));
  self.frame.ROLE_ICON:Show();
  self.frame.SPEC:Hide();
  if OmniPotent.OPTIONS.ICONS then self.frame.SPEC_ICON:Show();
  else self.frame.SPEC_ICON:Hide();
  end
end

function OmniPotentUnit:UpdateOrientation()
  self.frame.SPEC_ICON:ClearAllPoints();
  if self.parent.reverse then
    self.frame.SPEC_ICON:SetPoint('TOPLEFT', self.frame, 'TOPRIGHT', 1, 0);
  else
    self.frame.SPEC_ICON:SetPoint('TOPRIGHT', self.frame, 'TOPLEFT', -1, 0);
  end
end


OmniPotentAuras.lua
-- OmniPotentAuras
-- =====================================================================
-- Copyright (C) 2014 Lock of War, Developmental (Pty) Ltd
--

OmniPotentAuras = {
  parent=nil,
  max=9,
  count=1
};

OmniPotentAuras.__index = OmniPotentAuras;

function OmniPotentAuras:New(parent)
  local this = setmetatable({}, OmniPotentAuras);
  this.frames = setmetatable({}, nil);
  this.auras = setmetatable({}, nil);
  this.parent = parent;
  for i=1, this.max do
    this.frames[i] = CreateFrame('Button', parent.frame:GetName()..'Aura'..i, parent.frame, 'OmniPotentAuraTemplate');
    this.frames[i]:SetScript('OnUpdate', function(frame) this:OnUpdate(frame); end);
    this.frames[i]:Show();
  end
  return this;
end

function OmniPotentAuras:SetAura(count, frame, id, name, duration, expires, icon)
  if not self.auras[id] then
    self.auras[id] = name;
    frame.id = id;
    frame.spell = name;
    frame.update = GetTime();
    frame.duration = duration;
    frame.time = expires-GetTime();
    frame.icon = icon;
    frame.ICON:SetTexture(icon);
    frame:ClearAllPoints();
    if self.parent.parent.reverse then frame:SetPoint('TOPRIGHT', frame:GetParent(), 'TOPLEFT', -4-((count-1)*38), 0);
    else frame:SetPoint('TOPLEFT', frame:GetParent(), 'TOPRIGHT', 4+((count-1)*38), 0);
    end
    return 1;
  end
  return 0;
end

function OmniPotentAuras:UnsetAura(frame)
  frame.id = nil;
  frame.spell = nil;
  frame.update = GetTime();
  frame.duration = 0;
  frame.time = 0;
  frame.icon = nil;
  frame.ICON:SetTexture(nil);
  frame.expires:SetText('');
end

function OmniPotentAuras:UnitAura(unit)
  if OmniPotent.OPTIONS.AURAS then
    self.auras = table.wipe(self.auras);
    self.count = self:UpdatePlayerDebuff(1, unit);
    self.count = self:UpdateDebuff(self.count, unit);
    for i=self.count,self.max do
      self:UnsetAura(self.frames[i]);
    end
  end
end

function OmniPotentAuras:UpdatePlayerDebuff(count, unit)
  for i=1,40 do
    if count > self.max then break; end
    local name, rank, icon, stack, type, duration, expires, source, _, _, id = UnitDebuff(unit, i, 'PLAYER');
    if name and icon then
      count = count+self:SetAura(count, self.frames[count], id, name, duration, expires, icon);
    end
  end
  return count;
end

function OmniPotentAuras:UpdateDebuff(count, unit)
  for i=1,40 do
    if count > self.max then break; end
    local name, rank, icon, stack, type, duration, expires, source, _, _, id = UnitDebuff(unit, i);
    if name and icon then
      count = count+self:SetAura(count, self.frames[count], id, name, duration, expires, icon);
    end
  end
  return count;
end

function OmniPotentAuras:UpdatePositions()
  local count = 1;
  for i=1,self.max do
    if self.frames[i].id then
      self.frames[i]:ClearAllPoints();
      if self.parent.parent.reverse then self.frames[i]:SetPoint('TOPRIGHT', self.frames[i]:GetParent(), 'TOPLEFT', -4-((count-1)*38), 0);
      else self.frames[i]:SetPoint('TOPLEFT', self.frames[i]:GetParent(), 'TOPRIGHT', 4+((count-1)*38), 0);
      end
      count = count+1;
    end
  end
end

function OmniPotentAuras:Destroy()
  for i=1,self.max do
    self:UnsetAura(self.frames[i]);
  end
end

function OmniPotentAuras:UpdateExpires(frame)
  if frame.time then
    frame.time = tonumber(frame.time) - (GetTime()-frame.update);
    frame.time = math.floor((frame.time*10)+0.5)/10;
    if frame.duration == 0 then
      frame.expires:SetText('');
    elseif frame.time < 0 then
      self:UnsetAura(frame);
      self:UpdatePositions();
    elseif frame.time <= 60 then
      local time = frame.time > 0 and frame.time or '';
      frame.expires:SetText(time);
    else
      local msg, val = SecondsToTimeAbbrev(frame.time);
      frame.expires:SetText(format(msg, val));
    end
  end
end

function OmniPotentAuras:OnUpdate(frame)
  if frame.id then
    if GetTime()-frame.update >= 0.1 then
      self:UpdateExpires(frame);
      frame.ICON:SetTexture(frame.icon);
      frame.update = GetTime();
    end
  end
end

OmniPotent.lua
-- OmniPotent v1.2.2
-- =====================================================================
-- Copyright (C) 2014 Lock of War, Developmental (Pty) Ltd
--
-- This Work is provided under the Creative Commons
-- Attribution-NonCommercial-NoDerivatives 4.0 International Public License
--
-- Please send any bugs or feedback to omnipotent@lockofwar.com.
-- Debug /run print((select(4, GetBuildInfo())));

local WIDTH = GetScreenWidth()*UIParent:GetEffectiveScale();
local DEFAULT_OPTIONS = {
  VERSION=1.22,
  ENABLED=true,
  FRIENDLY=true,
  POSITION={
    HARMFUL={ 'RIGHT', nil, 'RIGHT', -250, 100 },
    HELPFUL={ 'LEFT', nil, 'LEFT', 250, 100 }
  },
  BORDERLESS=false,
  ICONS=true,
  SIZE=100,
  NAMING='Transmute',
  POWER=true,
  RANGE=true,
  TARGETED=true,
  AURAS=true
};

local FRIENDS = { 'Garrosh', 'Jaina', 'Uther', 'Anduin', 'Thrall' };
local ENEMIES = { 'Афила', 'Сэйбот', 'Яджун', 'Найнс', 'Айвен' };

OmniPotent = CreateFrame('Frame', 'OmniPotent', UIParent);

function OmniPotent:Load()
  self.active=false;
  self.version=DEFAULT_OPTIONS.VERSION;
  self.version_text='v1.2.2';
  self.frames={};
  self.player={};
  self.objectives=false;
  self:HelloWorld();
  self:Initialize();
  self:Options();
end

function OmniPotent:Initialize()
  self.frames.HELPFUL = OmniPotentGroup:New('HELPFUL', true);
  self.frames.HARMFUL = OmniPotentGroup:New('HARMFUL', false);
end

function OmniPotent:Activate()
  if self.OPTIONS.FRIENDLY then
    self.frames.HELPFUL:Activate();
  end
  self.frames.HARMFUL:Activate();
end

function OmniPotent:ZoneChanged()
  local active, battlefield = IsInInstance();
  if self.OPTIONS.ENABLED and not self.active and battlefield == 'arena' then
    self.active = active;
    self:DisableOptions();
    self:ObjectivesFrame(active);
    self:Activate();
  elseif not active then
    self.active = active;
    self:EnableOptions();
    self:ObjectivesFrame(active);
    self:Destroy();
  end
end

function OmniPotent:ObjectivesFrame(active)
  if ObjectiveTrackerFrame then
    if active and not ObjectiveTrackerFrame.collapsed then
      ObjectiveTracker_Collapse();
      self.objectives = true;
    elseif not active and self.objectives and ObjectiveTrackerFrame.collapsed then
      ObjectiveTracker_Expand();
      self.objectives = false;
    end
  end
end

function OmniPotent:PlayerLogin()
  self.player.NAME = UnitName('player');
  self.player.CLASS = select(2, UnitClass('player'));
  self.player.FACTION = GetBattlefieldArenaFaction();
end

function OmniPotent:HelloWorld()
  local message = '|cFF00FFFF <OmniPotent-'..self.version_text..'>|cFFFF0000 Be OP.|r Type /op for interface options.';
  ChatFrame1:AddMessage(message, 0, 0, 0, GetChatTypeIndex('SYSTEM'));
end

function OmniPotent:Options()
  self.OPTIONS = OMNIPOTENT_OPTIONS or nil;
  if not self.OPTIONS or self.OPTIONS.VERSION ~= self.version then
    self.OPTIONS = DEFAULT_OPTIONS;
  else
    for k, value in pairs(DEFAULT_OPTIONS) do
      if self.OPTIONS[k] == nil then
        self.OPTIONS[k] = value;
      end
    end
  end
  self:InitOptions();
end

function OmniPotent:InitOptions()
  self.options_frame = CreateFrame('Frame', 'OmniPotentOptions', UIParent, 'OmniPotentOptionsTemplate');
  self.options_frame.name = GetAddOnMetadata('OmniPotent', 'Title');
  self.options_frame.Title:SetText(string.upper('Arena/Skirmish'));
  self.options_frame.Subtitle:SetText('OmniPotent '..self.version_text);
  self.options_frame.okay = function() OmniPotent:SaveOptions(); end;
  self.options_frame.default = function() OmniPotent:DefaultOptions(); end;
  self.options_frame.cancel = function() OmniPotent:CancelOptions(); end;
  self.options_frame:SetScript('OnHide', function(self) OmniPotent:CloseOptions(); end);
  self.options_frame:SetScript('OnShow', function(self) OmniPotent:OpenOptions(); end);
  InterfaceOptions_AddCategory(self.options_frame);
  self:SetOptions(self.OPTIONS);
end

function OmniPotent:InitNamingOptions(default)
  UIDropDownMenu_Initialize(self.options_frame.Naming, function()
    for i, option in pairs({ 'Transmute', 'Transliterate', 'Ignore' }) do
      UIDropDownMenu_AddButton({ owner=self.options_frame.Naming, text=option, value=option, checked=nil, arg1=option, func=(function(_, value)
        UIDropDownMenu_ClearAll(self.options_frame.Naming);
        UIDropDownMenu_SetSelectedValue(self.options_frame.Naming, value);
        self:SetOption('naming', value);
      end)});
    end
  end);
  UIDropDownMenu_SetAnchor(self.options_frame.Naming, 16, 22, 'TOPLEFT', self.options_frame.Naming:GetName()..'Left', 'BOTTOMLEFT');
  UIDropDownMenu_JustifyText(self.options_frame.Naming, 'LEFT');
  UIDropDownMenu_SetSelectedValue(self.options_frame.Naming, default);
end

function OmniPotent:SetOptions(options)
  self.options_frame.Enabled:SetChecked(options.ENABLED);
  self.options_frame.Friendly:SetChecked(options.FRIENDLY);
  self.options_frame.Power:SetChecked(options.POWER);
  self.options_frame.Range:SetChecked(options.RANGE);
  self.options_frame.Targeted:SetChecked(options.TARGETED);
  self.options_frame.Auras:SetChecked(options.AURAS);
  self.options_frame.Borderless:SetChecked(options.BORDERLESS);
  self.options_frame.Icons:SetChecked(options.ICONS);
  self.options_frame.Size:SetValue(options.SIZE);
  self:InitNamingOptions(options.NAMING);
  if not InCombatLockdown() then
    self:SetOptionSize(options.SIZE);
    self:SetOptionPosition(options.POSITION);
  end
end

function OmniPotent:SaveOptions()
  OMNIPOTENT_OPTIONS = self.OPTIONS;
  OMNIPOTENT_OPTIONS.ENABLED = self.options_frame.Enabled:GetChecked();
  OMNIPOTENT_OPTIONS.FRIENDLY = self.options_frame.Friendly:GetChecked();
  OMNIPOTENT_OPTIONS.POWER = self.options_frame.Power:GetChecked();
  OMNIPOTENT_OPTIONS.RANGE = self.options_frame.Range:GetChecked();
  OMNIPOTENT_OPTIONS.TARGETED = self.options_frame.Targeted:GetChecked();
  OMNIPOTENT_OPTIONS.AURAS = self.options_frame.Auras:GetChecked();
  OMNIPOTENT_OPTIONS.BORDERLESS = self.options_frame.Borderless:GetChecked();
  OMNIPOTENT_OPTIONS.ICONS = self.options_frame.Icons:GetChecked();
  OMNIPOTENT_OPTIONS.SIZE = self.options_frame.Size:GetValue();
end

function OmniPotent:CancelOptions() OmniPotent:SetOptions(self.OPTIONS); end
function OmniPotent:DefaultOptions() self:SetOptions(DEFAULT_OPTIONS); end

function OmniPotent:DisableOptions()
  UIDropDownMenu_DisableDropDown(self.options_frame.Naming);
  self.options_frame.Enabled:Disable();
  self.options_frame.Friendly:Disable();
  self.options_frame.Power:Disable();
  self.options_frame.Auras:Disable();
  self.options_frame.Borderless:Disable();
  self.options_frame.Icons:Disable();
end

function OmniPotent:EnableOptions()
  UIDropDownMenu_EnableDropDown(self.options_frame.Naming);
  self.options_frame.Enabled:Enable();
  self.options_frame.Friendly:Enable();
  self.options_frame.Power:Enable();
  self.options_frame.Auras:Enable();
  self.options_frame.Borderless:Enable();
  self.options_frame.Icons:Enable();
end

function OmniPotent:SetOption(option, value)
  self.OPTIONS[string.upper(option)] = value;
  self:OpenOptions();
end

function OmniPotent:SetOptionPosition(position)
  self:SetOptionPositionHARMFUL(position.HARMFUL);
  self:SetOptionPositionHELPFUL(position.HELPFUL);
end

function OmniPotent:SetOptionPositionHARMFUL(position)
  self.frames.HARMFUL.frame:ClearAllPoints();
  local ok, point, relativeTo, relativePoint, x, y = pcall(unpack, position);
  if not ok then point, relativeTo, relativePoint, x, y = unpack(DEFAULT_OPTIONS.POSITION.HARMFUL);
  end
  self.frames.HARMFUL.frame:SetPoint(point, relativeTo, relativePoint, x, y);
  self.frames.HARMFUL:UpdateOrientation();
end

function OmniPotent:SetOptionPositionHELPFUL(position)
  self.frames.HELPFUL.frame:ClearAllPoints();
  local ok, point, relativeTo, relativePoint, x, y = pcall(unpack, position);
  if not ok then point, relativeTo, relativePoint, x, y = unpack(DEFAULT_OPTIONS.POSITION.HELPFUL);
  end
  self.frames.HELPFUL.frame:SetPoint(point, relativeTo, relativePoint, x, y);
  self.frames.HELPFUL:UpdateOrientation();
end

function OmniPotent:SetOptionSize(size)
  self.frames.HELPFUL.frame:SetScale(size/100);
  self.frames.HARMFUL.frame:SetScale(size/100);
end

function OmniPotent:OpenOptions()
  if not self.active then
    if self.OPTIONS.ENABLED then
      self:ObjectivesFrame(true);
      if self.OPTIONS.FRIENDLY then self.frames.HELPFUL:CreateStub(FRIENDS);
      end
      self.frames.HARMFUL:CreateStub(ENEMIES);
      if self.OPTIONS.BORDERLESS then self.options_frame.Icons:Show();
      else self.options_frame.Icons:Hide();
      end
    else
      self:Destroy();
    end
  end
end

function OmniPotent:CloseOptions()
  if not self.active then
    self:ObjectivesFrame(false);
    self.options_frame:Hide();
    self:Destroy();
  end
end

function OmniPotent:Destroy()
  self.frames.HARMFUL:Destroy();
  self.frames.HELPFUL:Destroy();
end

function OmniPotent:AddonLoaded(addon)
  if addon == 'OmniPotent' then
    self:Load();
  end
end

function OmniPotent:OnEvent(event, ...)
  if event == 'ADDON_LOADED' then self:AddonLoaded(...);
  elseif event == 'PLAYER_LOGIN' then self:PlayerLogin();
  elseif event == 'ACTIVE_TALENT_GROUP_CHANGED' then self:PlayerLogin();
  elseif event == 'PLAYER_ENTERING_WORLD' then self:ZoneChanged();
  elseif event == 'ZONE_CHANGED' then self:ZoneChanged();
  elseif event == 'ZONE_CHANGED_NEW_AREA' then self:ZoneChanged();
  elseif event == 'ZONE_CHANGED_INDOORS' then self:ZoneChanged();
  end
end

OmniPotent:SetScript('OnEvent', OmniPotent.OnEvent);

OmniPotent:RegisterEvent('PLAYER_ENTERING_WORLD');
OmniPotent:RegisterEvent('ACTIVE_TALENT_GROUP_CHANGED');
OmniPotent:RegisterEvent('ZONE_CHANGED');
OmniPotent:RegisterEvent('ZONE_CHANGED_NEW_AREA');
OmniPotent:RegisterEvent('ZONE_CHANGED_INDOORS');
OmniPotent:RegisterEvent('PLAYER_LOGIN');
OmniPotent:RegisterEvent('ADDON_LOADED');

SLASH_OMNIPOTENT1 = '/op';
SLASH_OMNIPOTENT2 = '/omni';
SLASH_OMNIPOTENT3 = '/omnipotent';
function SlashCmdList.OMNIPOTENT(cmd, box)
  InterfaceOptionsFrame_OpenToCategory(OmniPotent.options_frame);
  InterfaceOptionsFrame_OpenToCategory(OmniPotent.options_frame);
end